src/main/java/seedu/address/model/VersionedAddressBook.java

```diff
@@ -0,0 +1,90 @@
+package seedu.address.model;
+
+import java.util.ArrayList;
+
+import seedu.address.logic.commands.exceptions.CommandException;
+
+/**
+ * An address book that allows for undo and redo operations.
+ */
+public class VersionedAddressBook extends AddressBook {
+
+    public static final String UNDO_FAILURE = "No more commands left to undo";
+    public static final String REDO_FAILURE = "No more commands left to redo";
+
+    private final ArrayList<AddressBook> addressBookStateList;
```

Comment:
I like it! Good job. The implementation is clear. However, if storage might be an issue. In the future maybe limiting the number of states of storing the past 5 previous states can help with memory.